server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -143,12 +156,50 @@ public RoutingNodes(ClusterState clusterState, boolean readOnly) {
                 }
             }
         }
+        this.nodesToShards = new HashMap<>(nodesToShards.size());
         for (Map.Entry<String, LinkedHashMap<ShardId, ShardRouting>> entry : nodesToShards.entrySet()) {
             String nodeId = entry.getKey();
-            this.nodesToShards.put(nodeId, new RoutingNode(nodeId, clusterState.nodes().get(nodeId), entry.getValue()));
+            this.nodesToShards.put(nodeId, new RoutingNode(nodeId, discoveryNodes.get(nodeId), entry.getValue()));
         }
     }
 
+    private RoutingNodes(RoutingNodes routingNodes) {
+        this.readOnly = false;
+        this.nodesToShards = new HashMap<>(routingNodes.nodesToShards.size());
+        for (Map.Entry<String, RoutingNode> entry : routingNodes.nodesToShards.entrySet()) {
+            this.nodesToShards.put(entry.getKey(), entry.getValue().copy());
+        }
+        this.assignedShards = new HashMap<>(routingNodes.assignedShards.size());
+        for (Map.Entry<ShardId, List<ShardRouting>> entry : routingNodes.assignedShards.entrySet()) {
+            this.assignedShards.put(entry.getKey(), new ArrayList<>(entry.getValue()));
+        }
+        this.unassignedShards = routingNodes.unassignedShards.copyFor(this);
+
+        this.inactivePrimaryCount = routingNodes.inactivePrimaryCount;
+        this.inactiveShardCount = routingNodes.inactiveShardCount;
+        this.relocatingShards = routingNodes.relocatingShards;
+        this.activeShardCount = routingNodes.activeShardCount;
+        this.totalShardCount = routingNodes.totalShardCount;
+        this.attributeValuesByAttribute = new HashMap<>(routingNodes.attributeValuesByAttribute.size());
```

Comment:
This is a lazy cache of some immutable properties of the immutable set of nodes so it's never meaningfully mutated, so really we could just re-use the same map, or else instantiate it as an empty map. Also if it's only used in `AwarenessAllocationDecider` it should be empty on immutable instances anyway. But maybe it's nicer to explicitly copy it here anyway just for clarity.

Maybe we should move this to `DiscoveryNodes` so it doesn't get invalidated on every cluster state update too. Probably not a big deal, just thinking out loud here.